README.rst

```diff
@@ -0,0 +1,191 @@
+    ...
+    Xkey: 10a91397b7e74b8fb449a212147da89d ebbec6674baa4b2e9acdc1d8706c973e 943f8ea1a2f7400ebef9cd4b9cdc7692
+
+Inspecting the header reveals all the UUIDs of objects involved for rendering the HTML:
+
+* 10a91397b7e74b8fb449a212147da89d (the document itself)
+* ebbec6674baa4b2e9acdc1d8706c973e (image of a cute kitty)
+* 943f8ea1a2f7400ebef9cd4b9cdc7692 (image of an equally cute puppy)
+
+A second document exposes:
+
+.. code-block:: bash
+
+    $ curl -s -I -X GET http://localhost:8080/Plone/my-pets/my-cat-pixie
+
+    HTTP/1.1 200 OK
+    ...
+    Xkey: 5540249e9aa84020856454b001737775 ebbec6674baa4b2e9acdc1d8706c973e
+
+Now, after updating the image (with UUID ebbe5c...), all URLs that use this image are purged in Varnish.
+This does include of course all variants of the URLs of these two documents (potentially with query parameters like ``utm_campaign`` and such) and every scale of the updated image as well.
+
+How does it work? How to extend it?
+-----------------------------------
+
+During the publishing process all involved IDs (UUIDs and custom IDs) are collected (by subscribing to IPubAfterTraversal).
+
+Important are the adapters for IInvolvedIDs, which are responsible for collecting IDs for their given context.
+The base implementation looks for the UUIDs, but may be specialized for your custom content types.
+
+Apart from the adapter approach, there is the inline approach. You may call the following methods from collective.xkey.utils:
+
+* mark_involved_objects(request, objs, stoponfirst=False)
+* mark_involved(request, single_id)
+
+Whereas the first method uses internally the adapters for IInvolvedIDs for the given objects, the second one allows setting any arbitrary IDs.
+These methods combined might be used in your views, whenever a certain object or part is being rendered.
+
+Additionally, there is a utility browser view "mark_involved", that can be used in your template as follows:
+
+.. code-block:: xml
+
+    <tal:image tal:define="image python:context.get_image()" tal:condition="python: image">
+
+        <tal:mark-involved tal:define="involved image/@@mark_involved" />
+        <!-- put image rendering here -->
+
+    </tal:image>
+
+Alternatively, you can again set arbitrary IDs:
+
+.. code-block:: xml
+
+    <tal:mark-involved tal:define="involved context/@@mark_involved/my_custom_id" />
+
+After having collected all IDs a ITransform adapter puts the expected Xkey header in the HTTP response header.
+
+When Plone sends a purge request to the configured Cache Proxy, it sends additionally a specialized request for handling objects with tags.
+
+Installation
+------------
+
+The installation is two-fold. On the **Plone** side, you need to install collective.xkey by adding it to your buildout::
+
+    [buildout]
+
+    ...
+
+    eggs =
+        collective.xkey
+
+
+and then running ``bin/buildout``. Upon running up your instance, you may see the new header. No further setup for specific Plone sites is required.
```

Comment:
The sentence starting with "No further..." indicates that this is usable out of the box. Does the default implementation cover all objects involved in rendering a view? What about viewlets?

Reply:
The "setup" was meant no install add-on for plone site is needed. The default implementation covers alias URLs of objects. Will be clarified.